Taxonomy of failsafe levels

josephineSei opened the pull request:
closes https://github.com/SovereignCloudStack/standards/issues/527

markus-hentsch commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,89 @@
+Disk
+  A physical disc in a deployment.
+Node
+  A physical machine in a deployment.
```
```suggestion
  A physical machine in the infrastructure.
```

markus-hentsch commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,89 @@
+Node
+  A physical machine in a deployment.
+Cyber threat
+  Attacks on the cloud.
```
```suggestion
  Attacks on the infrastructure through the means of electronic access.
```

markus-hentsch commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,89 @@
+
+## Context
+
+Some standards in will talk about or require procedures to backup resources or have redundancy for resources.
```
I think a word is missing here.

markus-hentsch commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,89 @@
+
+Some standards in will talk about or require procedures to backup resources or have redundancy for resources.
+This decision record should discuss, which failure threats are CSPs facing and will group them into severel level.
+In consequence these levels should be used in standards talking about redundancy or failure-safety.
```
```suggestion
In consequence these levels should be used in standards concerning redundancy or failure-safety.
```

markus-hentsch commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,89 @@
+
+## Decision
+
+First there needs to be an overview about possible failure cases in deployments:
```
```suggestion
First there needs to be an overview about possible failure cases in infrastructures:
```

markus-hentsch commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,89 @@
+floating IP (abbr. FIP)
+  OpenStack resource, an IP that is usually reachable from the internet.
+Disk
+  A physical disc in a deployment.
```
I think "infrastructure" is a better fit and resembles the physical scope best. "Deployment" has some organizational level vibes to it imo. Also applies to any further comments concerning this term below.

markus-hentsch commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,89 @@
+| Node Outage | Medium to High | Data loss on node / (temporary) loss of functionality and connectivity of node (impact depends on type of node)  |
+| Rack Outage | Medium | similar to Disk Failure and Node Outage |
+| Power Outage (Data Center supply)  | Medium | potential data loss, temporary loss of functionality and connectivity of node (impact depends on type of node)  |
+| Fire | Medium | permanently Disk and Node loss in the affected zone |
```
```suggestion
| Fire | Medium | permanent Disk and Node loss in the affected zone |
```

markus-hentsch commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,89 @@
+| Rack Outage | Medium | similar to Disk Failure and Node Outage |
+| Power Outage (Data Center supply)  | Medium | potential data loss, temporary loss of functionality and connectivity of node (impact depends on type of node)  |
+| Fire | Medium | permanently Disk and Node loss in the affected zone |
+| Flood | Low | permanently Disk and Node loss in the affected zone |
```
```suggestion
| Flood | Low | permanent Disk and Node loss in the affected zone |
```

markus-hentsch commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,89 @@
+| Power Outage (Data Center supply)  | Medium | potential data loss, temporary loss of functionality and connectivity of node (impact depends on type of node)  |
+| Fire | Medium | permanently Disk and Node loss in the affected zone |
+| Flood | Low | permanently Disk and Node loss in the affected zone |
+| Earthquake | Very Low | permanently Disk and Node loss in the affected zone |
```
```suggestion
| Earthquake | Very Low | permanent Disk and Node loss in the affected zone |
```

markus-hentsch commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,89 @@
+| Fire | Medium | permanently Disk and Node loss in the affected zone |
+| Flood | Low | permanently Disk and Node loss in the affected zone |
+| Earthquake | Very Low | permanently Disk and Node loss in the affected zone |
+| Storm/Tornado | Low | permanently Disk and Node loss in the affected fire zone |
```
```suggestion
| Storm/Tornado | Low | permanent Disk and Node loss in the affected fire zone |
```

markus-hentsch commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,89 @@
+| Flood | Low | permanently Disk and Node loss in the affected zone |
+| Earthquake | Very Low | permanently Disk and Node loss in the affected zone |
+| Storm/Tornado | Low | permanently Disk and Node loss in the affected fire zone |
+| Cyber threat  | High | permanently loss of data on affected Disk and Node  |
```
```suggestion
| Cyber threat  | High | permanent loss or compromise of data on affected Disk and Node  |
```

markus-hentsch commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,89 @@
+| Storm/Tornado | Low | permanently Disk and Node loss in the affected fire zone |
+| Cyber threat  | High | permanently loss of data on affected Disk and Node  |
+
+These failure case can result in temporary (T) or permanent (P) loss of the resource or data within.
```
```suggestion
These failure cases can result in temporary (T) or permanent (P) loss of the resource or data within.
```

markus-hentsch commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,89 @@
+
+These failure case can result in temporary (T) or permanent (P) loss of the resource or data within.
+Additionally there are a lot of resources in IaaS alone that are more or less affected by these Failure Cases.
+The following table shows the affection without considering any redundancy or failure saftey being in use:
```
```suggestion
The following table shows the impact when no redundancy or failure safety measure is in place:
```

Looking into the dictionary, I don't think "affection" has the meaning you intended here ;)

markus-hentsch commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,89 @@
+| network connectivity (materialization) | | T (if on node) | T/P | T | P (T if lucky) | T/P |
+| floating IP | P (if on disk) | T (if on node) | T/P | T | P (T if lucky) | T/P |
+
+For some cases there are only temporary unavailabilites and clouds do have certain workflows to avoid data loss, like redundancy in storagy backends and databases.
```
```suggestion
For some cases, this only results in temporary unavailabilities and cloud infrastructures usually have certain mechanisms in place to avoid data loss, like redundancy in storage backends and databases.
```

markus-hentsch commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,89 @@
+| floating IP | P (if on disk) | T (if on node) | T/P | T | P (T if lucky) | T/P |
+
+For some cases there are only temporary unavailabilites and clouds do have certain workflows to avoid data loss, like redundancy in storagy backends and databases.
+So some of these outages are more easy to solve than others.
```
```suggestion
So some of these outages are easier to mitigate than others.
```

markus-hentsch commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,89 @@
+
+For some cases there are only temporary unavailabilites and clouds do have certain workflows to avoid data loss, like redundancy in storagy backends and databases.
+So some of these outages are more easy to solve than others.
+A possible way to group the failure cases into levels considering the matrix of affection would be:
```
```suggestion
A possible way to classify the failure cases into levels considering the matrix of impact would be:
```

(re "affection" - see above)

markus-hentsch commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,89 @@
+So some of these outages are more easy to solve than others.
+A possible way to group the failure cases into levels considering the matrix of affection would be:
+
+| Level/Class | level of affection | Use Cases |
```
```suggestion
| Level/Class | level of impact | Use Cases |
```

(re "affection" - see above)

markus-hentsch commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,89 @@
+| Level/Class | level of affection | Use Cases |
+|---|---|-----|
+| 1. Level | single volumes, VMs... | Disk Failure, Node outage, (maybe rack outage) |
+| 2. Level | number of resources, most of the time recoverable | Rack outage, (Fire), (Power outage when different power supplies exist) |
```
```suggestion
| 2. Level | limited number of resources, most of the time recoverable | Rack outage, (Fire), (Power outage when different power supplies exist) |
```

markus-hentsch commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,89 @@
+
+| Level/Class | level of affection | Use Cases |
+|---|---|-----|
+| 1. Level | single volumes, VMs... | Disk Failure, Node outage, (maybe rack outage) |
```
```suggestion
| 1. Level | individual volumes, VMs... | Disk Failure, Node outage, (maybe rack outage) |
```

markus-hentsch commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,89 @@
+| 1. Level | single volumes, VMs... | Disk Failure, Node outage, (maybe rack outage) |
+| 2. Level | number of resources, most of the time recoverable | Rack outage, (Fire), (Power outage when different power supplies exist) |
+| 3. Level | lots of resources / user data + potentially not recoverable | Fire, Earthquake, Storm/Tornado, Power Outage |
+| 4. Level | complete deployment, not recoverable | Flood, Fire |
```
```suggestion
| 4. Level | entire infrastructure, not recoverable | Flood, Fire |
```

markus-hentsch commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,89 @@
+| 3. Level | lots of resources / user data + potentially not recoverable | Fire, Earthquake, Storm/Tornado, Power Outage |
+| 4. Level | complete deployment, not recoverable | Flood, Fire |
+
+Unfortunately something similar does not seem to exist right now.
```
```suggestion
Based on our research, no similar standardized classification scheme seems to exist currently.
Thus, this decision record establishes its own.
```

markus-hentsch commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,89 @@
+
+## Consequences
+
+Using the definition of Levels throughout all SCS standards would allow readers to know up to which Level certain procedures or aspects of resources (e.g. volume types or a backend requiring redundancy) would protect their data.
```
```suggestion
Using the definition of levels established in this decision record throughout all SCS standards would allow readers to understand up to which level certain procedures or aspects of resources (e.g. volume types or a backend requiring redundancy) would protect their data and/or resource availability.
```

anjastrunk commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,90 @@
+This decision record aims to define different levels of failure-safety.
+These levels can then be used in standards to clearly set the scope that certain procedures in e.g. OpenStack offer.
+
+## Terminology
+
+Image
+  OpenStack resource, virtual machine images usually residing in a network storage backend.
+Volume
+  OpenStack resource, virtual drive which usually resides in a network storage backend.
+Virtual Machine (abbr. VM)
+  IaaS resource, also called server, executes workloads from users.
+Secret
+  OpenStack ressource, cryptographic asset stored in the Key Manager (e.g. Barbican).
+Key Encryption Key (abbr. KEK)
+  OpenStack resource, used to encrypt other keys to be able to store them encrypted in a database.
+Floating IP (abbr. FIP)
+  OpenStack resource, an IP that is usually routed and accessible from external networks.
+Disk
+  A physical disk drive (e.g. HDD, SSD) in the infrastructure.
+Node
+  A physical machine in the infrastructure.
+Cyber threat
+  Attacks on the infrastructure through the means of electronic access.
+
```
I know, it's bit picky and I like to have the same look and feel in our standards and decision records. I would display terminology as table and call it glossary.

```suggestion
## Glossary
| Term              | Explanation                                                                                                                                                |
| ----------------- | ---------------------------------------------------------------------------------------------------------------------------------------- |
| Image           | OpenStack resource, virtual machine images usually residing in a network storage backend.                  |  
Volume           | OpenStack resource, virtual drive which usually resides in a network storage backend.                           | 
| Virtual Machine (abbr. VM) |   IaaS resource, also called server, executes workloads from users.                                   | 
| Secret |   OpenStack ressource, cryptographic asset stored in the Key Manager (e.g. Barbican).                                    | 
| Key Encryption Key (abbr. KEK) |  OpenStack resource, used to encrypt other keys to be able to store them encrypted in a database.                                                                              | 
| Floating IP (abbr. FIP) |   OpenStack resource, an IP that is usually routed and accessible from external networks.       | 
| Disk | A physical disk drive (e.g. HDD, SSD) in the infrastructure.                                                                                    | 
| Node | A physical machine in the infrastructure.                                                                                                               | 
| Cyber threat | Attacks on the infrastructure through the means of electronic access.                                                       | 

```

josephineSei replied:
I took Markus table from the PR you mentioned above and adjusted it to comply.

For the title: I took the example DR from the standards repository. Maybe we should change the title from Terminology to Glossary there too.

anjastrunk commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,90 @@
+Volume
+  OpenStack resource, virtual drive which usually resides in a network storage backend.
+Virtual Machine (abbr. VM)
+  IaaS resource, also called server, executes workloads from users.
```
Why do we distinguish between IaaS resources and OpenStack resources? It there any added-value to categorize Virtual Machine as "IaaS Resource"

anjastrunk commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,90 @@
+
+## Context
+
+Some standards provided by the SCS will talk about or require procedures to backup resources or have redundancy for resources.
```
```suggestion
Some standards provided by the SCS project will talk about or require procedures to backup resources or have redundancy for resources.
```

anjastrunk commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,86 @@
+| Storm/Tornado | Low | permanent Disk and Node loss in the affected fire zone |
+| Cyber threat  | High | permanent loss or compromise of data on affected Disk and Node  |
+
+These failure cases can result in temporary (T) or permanent (P) loss of the resource or data within.
```
I like this table. It is a very good starting point to think about backup and failure strategies. IMO, its content is not in a final state. As well as the following paragraphs. They look a bit incomplete. We should take an other round on brainstorming to take OpenStack services and k8s resources and services into account as well. 

At the end, I like to see an overview on how use cases defined in upper table effect availability of all OpenStack and k8s resources and services. This overview MUST include all fail-safe strategies SCS project requires together with the use case they prevent. Maybe this overview would a separate standard or references other standards. We have to discuss this.

anjastrunk reviewed (changes requested):
IMO, this DR is not in a final state and we should go for  another round on discussion

josephineSei commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -12,30 +12,26 @@ When talking about redundancy and backups in the context of cloud infrastructure
+
+| Term               | Explanation                                                                                                                              |
+| ------------------ | ---------------------------------------------------------------------------------------------------------------------------------------- |
+| Virtual Machine    | Equals the `server` resource in Nova.                                                                                                    |
```
That is a good point. We definitely need to separate between VMs and Ironic nodes.

josephineSei commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,101 @@
+
+:::
+
+| Level/Class | Probability | Failure Causes | loss in IaaS | User Hints |
```
I am still thinking about the "user hints" column. Putting it next to the other columns is good from some perspectives, as it can be read: I want to achieve 2. Level of failuresafeness, which can be triggered by these failure causes that will result in these losses on IaaS level, so I can do, what is shown in the user hints.
But we wanted the classification not for examples for users, but mainly as a definiton for standards, so maybe we should not reference those standards here.
We could rather use an extra table with example actions(standards, "user has to to things",..) for each level/class or maybe this should rather not be in a decision record, but rather in a guide or so.

anjastrunk replied:
I agree with you, @josephineSei. Linking SCS standards in this table may cause a huge synchronization effort. We always have to update this DR, if referenced standards change. I appreciate the column user hints, but would limit to a textual explanation.  See my suggestions below...

anjastrunk commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,101 @@
+| (Volume) Snapshot  | Thinly-provisioned copy-on-write snapshots of volumes. Stored in the same Cinder storage backend as volumes.                             |
+| Volume Type        | Attribute of volumes determining storage details of a volume such as backend location or whether the volume will be encrypted.           |
+| (Barbican) Secret  | IaaS resource storing cryptographic assets such as encryption keys. Managed by the Barbican service.                                     |
+| Key Encryption Key | IaaS resource, used to encrypt other keys to be able to store them encrypted in a database.                                              |
```
I a key encryption key really an IaaS resource? I thought key encryption keys are stored in configuration files and if this is the case, it is a configuration setting.

josephineSei replied:
The Glossary is for all maybe unknown phrases to be described. As this standard also concerns Key Encryption Keys, it should be noted in the glossary

anjastrunk commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,101 @@
+| Key Encryption Key | IaaS resource, used to encrypt other keys to be able to store them encrypted in a database.                                              |
+| Floating IP        | IaaS resource, an IP that is usually routed and accessible from external networks.                                                       |
+| Disk               | A physical disk drive (e.g. HDD, SSD) in the infrastructure.                                                                             |
+| Node               | A physical machine in the infrastructure.                                                                                                |
```
If we list hardware resources, does it make sense to mention network node as an additional main single-point-of-failure?

```suggestion
| Node               | A physical machine in the infrastructure providing computational and/or storage capabilities.                                                                                                | Network node | A physical device providing network connectivity, such as switch or router. 
```

anjastrunk commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,101 @@
+
+| Level/Class | Probability | Failure Causes | loss in IaaS | User Hints |
+|---|---|---|-----|-----|
+| 1. Level | Very High | small Hardware or Software Failures (e.g. Disk/Node Failure, Software Bug,...) | individual volumes, VMs... | [volume replication](https://docs.scs.community/standards/scs-0114-v1-volume-type-standard) |
```
```suggestion
| 1. Level | Very High | small Hardware or Software Failures (e.g. Disk/Node Failure, Software Bug,...) | individual volumes, VMs... | CSP MUST use volume backend, which is replicated OR users MUST backup volumes themselves and place them on an other host. |
```

anjastrunk commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,101 @@
+| 1. Level | Very High | small Hardware or Software Failures (e.g. Disk/Node Failure, Software Bug,...) | individual volumes, VMs... | [volume replication](https://docs.scs.community/standards/scs-0114-v1-volume-type-standard) |
+| 2. Level | High | important Hardware or Software Failures (e.g. Rack outage, small Fire, Power outage, ...) | limited number of resources, sometimes recoverable | [volume backups](https://github.com/SovereignCloudStack/standards/pull/567) |
+| 3. Level | Medium | small catastrophes or major Failures (e.g. fire, regional Power Outage, orchestrated cyber attacks,...) | lots of resources / user data + potentially not recoverable | Availability Zones, user responsibility |
+| 4. Level | Low | whole deployment loss (e.g. natural desaster,...) | entire infrastructure, not recoverable | user responsibility |
```
```suggestion
| 4. Level | Low | whole deployment loss (e.g. natural disaster,...) | entire infrastructure, not recoverable | CSP is able to save user from such catastrophes. User is responsibility for saving his data from natural disasters. |
```

anjastrunk commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,101 @@
+|---|---|---|-----|-----|
+| 1. Level | Very High | small Hardware or Software Failures (e.g. Disk/Node Failure, Software Bug,...) | individual volumes, VMs... | [volume replication](https://docs.scs.community/standards/scs-0114-v1-volume-type-standard) |
+| 2. Level | High | important Hardware or Software Failures (e.g. Rack outage, small Fire, Power outage, ...) | limited number of resources, sometimes recoverable | [volume backups](https://github.com/SovereignCloudStack/standards/pull/567) |
+| 3. Level | Medium | small catastrophes or major Failures (e.g. fire, regional Power Outage, orchestrated cyber attacks,...) | lots of resources / user data + potentially not recoverable | Availability Zones, user responsibility |
```
```suggestion
| 3. Level | Medium | small catastrophes or major Failures (e.g. fire, regional Power Outage, orchestrated cyber attacks,...) | lots of resources / user data + potentially not recoverable | CPS SHOULD operate hardware in dedicated  Availability Zones. User SHOULD backup his data, hiself.  |
```

anjastrunk commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,101 @@
+| 4. Level | Low | whole deployment loss (e.g. natural desaster,...) | entire infrastructure, not recoverable | user responsibility |
+
+Based on our research, no similar standardized classification scheme seems to exist currently.
+Something close but also very detailed can be found in [this (german)](https://www.bsi.bund.de/SharedDocs/Downloads/DE/BSI/Grundschutz/BSI_Standards/standard_200_3.pdf?__blob=publicationFile&v=2) from the BSI.
```
I think BSI is known to non Germans. I suggest to use English translation

```suggestion
Something close but also very detailed can be found in [this (german)](https://www.bsi.bund.de/SharedDocs/Downloads/DE/BSI/Grundschutz/BSI_Standards/standard_200_3.pdf?__blob=publicationFile&v=2) from the German Federal Office for Information Security.
```

anjastrunk commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,101 @@
+
+Based on our research, no similar standardized classification scheme seems to exist currently.
+Something close but also very detailed can be found in [this (german)](https://www.bsi.bund.de/SharedDocs/Downloads/DE/BSI/Grundschutz/BSI_Standards/standard_200_3.pdf?__blob=publicationFile&v=2) from the BSI.
+As we want to focus on IaaS resources and also have an easily understandable structure that can be applied in standards covering replication, redundancy and backups, this document is too detailed.
```
This DR should focus both IaaS and K8S resources. For now it covers IaaS resources only, but it should be extended by K8S resources,  soon. 

```suggestion
As we want to focus on IaaS and K8s resources and also have an easily understandable structure that can be applied in standards covering replication, redundancy and backups, this document is too detailed.
```

anjastrunk reviewed (approved):
Suggested some minor changes, but do not want to block PR.

anjastrunk commented:
@martinmo @cah-hbaum  @tonifinger Please update taxonomy with respect to kubernetes.

josephineSei commented:
@cah-hbaum , @martinmo We want to also include the KaaS level into this document. Especially having a second table, like the last one, that looks at the failure levels from a KaaS perpective and add the loss for KaaS resources and user hints on what to do.

We can also discuss this topic again, you can just ping me if you have any questions.

cah-hbaum commented:
> @cah-hbaum , @martinmo We want to also include the KaaS level into this document. Especially having a second table, like the last one, that looks at the failure levels from a KaaS perpective and add the loss for KaaS resources and user hints on what to do.
> 
> We can also discuss this topic again, you can just ping me if you have any questions.

I added some failure cases for K8s as a new table.
I'm gonna add some user hints for how to deal with these problems next.
@martinmo if you have some additional things to include, just make a commit (or fixup)

martinmo commented:
I like the idea of this document and see why it's necessary :+1: Based on @cah-hbaum's input I took the freedom to add some commits:

* For example, because of the addition of infos for the KaaS layer, I tried to structure the impacts a little bit better in this regard.
* The glossary was mainly for IaaS and I added corresponding info for the KaaS layer.
* Furthermore I resolved an IaaS vs KaaS ambiguity with respect to the the term *node* by simply replacing this term with *host* for IaaS, such that *node* can be used exclusively in the K8s context (**to be discussed**).

@josephineSei I am not finished yet, unfortunaly. Therefore I marked the PR as draft again and also assigned myself, I hope that is okay. IMHO this is a very difficult topic. Especially, I was stuck at the impacts (temporary/permanent loss) and can't conclude how this is best done. I tend to mark everything which is "stateless" with a "T". (I'll approach you to discuss this in person.)

I am also confused by some details: for example, we assume a node loss implicates a disk loss. This is not always the case, e.g., if the disk is still fine and only the motherboard had a malfunction, right?

josephineSei commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,182 @@
+| Kubelet software fault                       | Medium      | Crashing kubelet cannot start new pods on the node / kubelet might delete the pods or not / node marked unhealthy / replication controllers start new pods elsewhere |
+| Cluster operator error                       | Medium      | Loss of pods, services, etc. / lost of apiserver backing store / users unable to read API                                                                            |
+| Failure of multiple nodes or underlying DB   | Low         | Possible loss of all data depending on the amount of nodes lost compared to the cluster size, otherwise costly rebuild                                               |
+
```
I think we need another section here to better structure the document.
```suggestion

### Impact of the failure cases

```

josephineSei commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,182 @@
+|PVC|P| | | | | |P|
+|API Server|T| | | | | |T/P|
+
+For some cases, this only results in temporary unavailability and cloud infrastructures usually have certain mechanisms in place to avoid data loss, like redundancy in storage backends and databases.
```
These sentences are IaaS-specific I think. They belong under the table "Impact on OpenStack Resources (IaaS)"

josephineSei commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,182 @@
+| 2. Level | High | important Hardware or Software Failures (e.g. Rack outage, small Fire, Power outage, ...) | limited number of resources, sometimes recoverable | CSPs MUST operate replicas for important components (e.g. replicated volume back-end, uninterruptible power supply, ...) OR users MUST backup their data themselves and place it on an other host. |
+| 3. Level | Medium | small catastrophes or major Failures (e.g. fire, regional Power Outage, orchestrated cyber attacks,...) | lots of resources / user data + potentially not recoverable | CPSs SHOULD operate hardware in dedicated Availability Zones. Users SHOULD backup their data, themself. |
+| 4. Level | Low | whole deployment loss (e.g. natural disaster,...) | entire infrastructure, not recoverable | CSPs may not be able to save user data from such catastrophes. Users are responsible for saving their data from natural disasters. |
+
```
A table like this for KaaS is still missing, I think, you are on it @martinmo ?

martinmo replied:
My goal was to end up with common levels for both layers.

josephineSei replied:
yes the levels should be the same.

To have two tables is just for better readability and structure (so you can directly go to either IaaS or KaaS impacts and hints.)

josephineSei commented:
> I like the idea of this document and see why it's necessary 👍 Based on @cah-hbaum's input I took the freedom to add some commits:
> 
>     * For example, because of the addition of infos for the KaaS layer, I tried to structure the impacts a little bit better in this regard.
> 
>     * The glossary was mainly for IaaS and I added corresponding info for the KaaS layer.
> 
>     * Furthermore I resolved an IaaS vs KaaS ambiguity with respect to the the term _node_ by simply replacing this term with _host_ for IaaS, such that _node_ can be used exclusively in the K8s context (**to be discussed**).
> 
> 
> @josephineSei I am not finished yet, unfortunaly. Therefore I marked the PR as draft again and also assigned myself, I hope that is okay. IMHO this is a very difficult topic. Especially, I was stuck at the impacts (temporary/permanent loss) and can't conclude how this is best done. I tend to mark everything which is "stateless" with a "T". (I'll approach you to discuss this in person.)
> 
> I am also confused by some details: for example, we assume a node loss implicates a disk loss. This is not always the case, e.g., if the disk is still fine and only the motherboard had a malfunction, right?

Temporary and Permanent refers to data, so I think you are correct marking stateless resources, as long as their "building plan" is still available and they can be brought up quickly.

If a node is lost, then yes the disk may be still working, but it is not available. There needs to be physical work done to check that. That requires such a large amount of time, that at least for that timeframe the disk ans all data on it is also lost. If a VM is e.g. working with volumes (has a boot-volume) it is way easier to use a snapshot or internal ceph replication to bring up a new VM.

josephineSei commented:
As this will also be needed for the Availability Zone Standard, I am working on this MR, trying to understand the new structure and applying some of my suggestions.

josephineSei commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -13,11 +13,127 @@ There does exist very detailed lists of risks and what consequences there are fo
+* **Level 3**: CSPs SHOULD operate hardware in dedicated regions.
+* **Level 4**: Depending on the regions, CSPs may not be able to save user data from such catastrophes.
+
+More specific guidance on what these levels mean on the IaaS and KaaS layers will be provided in the sections
```
I integrated a table to show both consequences for CSPs and users. We may discuss again, because you mentioned to already user AZs in Level 2, while this is imho more a level 3 consequence.

josephineSei commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -13,11 +13,127 @@ There does exist very detailed lists of risks and what consequences there are fo
 
+<!--
+TODO: goals vs. non-goals
+TODO: Resolve confusion between high availability vs. disaster recovery vs. redundancy vs. backups
```
I moved them down and started writing about this a little bit.

josephineSei commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,264 @@
+In general, the lowest, **level 1**, describes isolated/local failures which can occur very frequently, whereas
+the highest, **level 4**, describes relatively unlikely failures that impact a whole or even multiple datacenter(s):
+
+| Level | Probability | Impact                | Examples |
```
I'm still not sure, if we should start with this. I know there are two was to present a decision. Mine was: 1. explain everything that leads to that decision and 2. the decision.
Now it is the other way around: 1. decision 2. why did we come to that decision.

This is just my personal style: I like having the explanation before. But we can have a discussion about this, maybe for this long long topic, people do not want to read through everything and come to the most important table quickly.

josephineSei commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,264 @@
+
+Using the definition of levels established in this decision record throughout all SCS standards would allow readers to understand up to which level certain procedures or aspects of resources (e.g. volume types or a backend requiring redundancy) would protect their data and/or resource availability.
+
+### General Terms
```
Were these tables the glossary before? Because it looks a little bit like it and would explain anjas comment and my answer. 

I don't get completely what you want to achieve with having these down here? what should this whole part of the decision record about?

josephineSei commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,317 @@
+### Affected Resources
+
+#### IaaS Layer (OpenStack Resources)
+
```
Affect resource table: we should add  from which level/class of failuresafeness, these resources are affected.

josephineSei commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,317 @@
+
+TODO: why should we do that?
+
+#### Impact on OpenStack Resources (IaaS layer)
```
Put into another document (like implementation record): Impact record.

josephineSei commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,317 @@
+
+A similar overview can be provided for Kubernetes infrastructures. These also include the things mentioned for infrastructure failure scenario, since a Kubernetes cluster
+would most likely be deployed on top of this infrastructure or face similar problems on a bare-metal installation.
+Part of this list comes directly from the official [Kubernetes docs](https://kubernetes.io/docs/tasks/debug/debug-cluster/).
```
this should be up with "affected resources"

josephineSei commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,317 @@
+|---|-----|-----|
+| 1. Level | CSPs MUST operate replicas for important components (e.g. replicated volume back-end, uninterruptible power supply, ...). | Users SHOULD backup their data themself and place it on an other host. |
+| 2. Level | CSPs MUST operate replicas for important components (e.g. replicated volume back-end, uninterruptible power supply, ...) | Users MUST backup their data themselves and place it on an other host. |
+| 3. Level | CPSs SHOULD operate hardware in dedicated Availability Zones. | Users SHOULD backup their data, in different AZs or even other deployments. |
```
CSP not CPS ;)

josephineSei commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,317 @@
+| Level | consequences for CSPs | consequences for Users |
+|---|-----|-----|
+| 1. Level | CSPs MUST operate replicas for important components (e.g. replicated volume back-end, uninterruptible power supply, ...). | Users SHOULD backup their data themself and place it on an other host. |
+| 2. Level | CSPs MUST operate replicas for important components (e.g. replicated volume back-end, uninterruptible power supply, ...) | Users MUST backup their data themselves and place it on an other host. |
```
```suggestion
| 2. Level | CSPs MUST have redundancy for important components (e.g. replicated volume back-end, uninterruptible power supply, ...) | Users MUST backup their data themselves and place it on an other host. |
```

josephineSei commented:
@martinmo and I discussed the structure of the document, and I will update it accordingly.

josephineSei commented:
I rephrased some paragraphs, added new and added an "implementation note" that I would like to keep, because it contains the cross-table of impact. I also added tables and looked again over the probability of failure scenarios.
@bitkeks could you check the security part of this documents, which are the failure cases and their probabilities?

I think the IaaS part is in good state now, so I would like someone from the Container team to add their knowledge about containers to this document, @tonifinger could you do this?

michal-gubricky commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,254 @@
+
+| Resource(s)        | Explanation                                                                                                                              | Affected by Level |
+| ------------------ | ---------------------------------------------------------------------------------------------------------------------------------------- | ----------------- |
+| Pod                | Kubernetes object that represents a workload to be executed, consisting of one or more containers.                                       | ???               |
```
I want to ask if the `Affected by Level` column will be filled in, or are the question marks intended to remain there?

josephineSei replied:
It should be filled. But that is the point I need help with, as I have up to now mostly worked with the IaaS Layer, not the KaaS Layer.

jschoone replied:
Hi @michal-gubricky and @josephineSei,
@DEiselt and I removed the questions marks now to fill in the levels

michal-gubricky replied:
Thanks, I will take a look at it.

michal-gubricky replied:
The changes look good to me. I will resolve this thread.

josephineSei commented:
Hey @jschoone or @tonifinger or @DEiselt could someone of you help with the KaaS part of this document (or do you know someone who can?) until the end of the week?

DEiselt commented:
> Hey @jschoone or @tonifinger or @DEiselt could someone of you help with the KaaS part of this document (or do you know someone who can?) until the end of the week?

Hi Josephine, Jan and i will meet today to discuss this and provide feedback :)

anjastrunk commented:
@mbuechse @kgube @michal-gubricky @jschoone: Please review, request changes and/or approve. We want to merge this PR as soon as possible.

kgube commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,254 @@
+| Compute             | A generic name for the IaaS service, that manages virtual machines (e.g. Nova in OpenStack).                                             |
+| Network             | A generic name for the IaaS service, that manages network resources (e.g. Neutron in OpenStack).                                         |
+| Storage             | A generic name for the IaaS service, that manages the storage backends and virtual devices (e.g. Cinder in OpenStack).                   |
+| RTO                 | Recovery Time Objective.                                                                                                                 |
```
```suggestion
| RTO                 | Recovery Time Objective, the time within which a service must be restored.                                                               |
```

bitkeks replied:
"the acceptable time"

josephineSei replied:
I wrote in other words, because to me your sentence was confusing. Please check, if that is, what you wanted to say.

kgube replied:
yeah, that works for me

kgube commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,254 @@
+| High Availability  | Refers to the availability of resources over an extended period of time unaffected by smaller hardware issues. E.g. achievable through having several instances of resources.      |
+| Disaster Recovery  | Measures taken after an incident to recover data, IaaS resource and maybe even physical resources.                                                                                 |
+| Redundancy         | Having more than one (or two) instances of each resource, to be able to switch to the second resource (could also be a data mirror) in case of a failure.                          |
+| Backup             | A specific copy of user data, that presents all data points at a givne time. Usually managed by users themself, read only and never stored in the same place as the original data. |
```
```suggestion
| Backup             | A specific copy of user data, that presents all data points at a given time. Usually managed by users themself, read only and never stored in the same place as the original data. |
```

kgube commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,254 @@
+
+| Failure Scenario | Probability | Consequences | Failsafe Level Coverage |
+|----|-----|----|----|
+| Software bug (major) | Low to Medium | permanent loss or compromise of data that trigger the bug up to data on the whole physical machine | L3 |
```
As you write below:
```suggestion
| Software bug (major) | Low to Medium | permanent loss or compromise of data that trigger the bug up to data on the whole deployment | L3 |
```

kgube commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,254 @@
+Instead every deployment needs to be prepared for an attack all the time, e.g. through security updates.
+The severity of Cyber attacks can also vary broadly: from denial-of-service attacks, which should only be a temporary issue, up until coordinated attacks to steal or destroy data, which could also affect a whole deployment.
+The easier an attack is, the more frequently it will be used by various persons and organizations up to be just daily business.
+Major attacks are often orchestrated and require speicif knowledge e.g. of Day-0 Bugs or the attacked infrastructure.
```
```suggestion
Major attacks are often orchestrated and require specific knowledge e.g. of Day-0 Bugs or the attacked infrastructure.
```

kgube reviewed (changes requested):
Some small fixes and suggestions

bitkeks commented on Standards/scs-XXXX-vN-taxonomy-of-failsafe-levels.md:
```diff
@@ -0,0 +1,254 @@
+It should be taken into consideration that the RTO for IaaS and KaaS means to make user data available again through measures within the infrastructure.
+But this will not be effective, when there is no backup of the user data or a redundancy of it already in place.
+The different failsafe levels, measures and impacts will lead to very different RTOs.
+For example a storage disk that has a failure will result in an RTO of 0 seconds, when the storage backend uses internal replication and still has two replicas of the user data.
+While in the worst case of a natural disaster, most likely a severe fire, the whole deployment will be lost and if there were no off-site backups done by users there will be no RTO, because the data cannot be recovered anymore.
```
Maybe the usage of RTO was meant differently by the writer here. In general, RTO is a set time frame (based on risk analysis or BCM), not a measurement of things that happen. What more likely fits here in this paragraph is **MTTR**, mean time to repair.

josephineSei replied:
I rephrased the paragraph, could you look over it again please?

bitkeks replied:
:+1:

bitkeks reviewed (approved):
LGTM, left one comment regarding the usage of RTO.

Good approach of defining the levels, then the potential influences and the many components that are affected. Mapping this all to the SCS technology parts will help CSPs and users to identify those parts of their instance they need to have a look at more specifically.

Of course this document will probably evolve with more empirical data - you could for example discuss, if river floods in Germany are really "very low" probability :smiley: 

Very important also to highlight the need for a CSP-internal risk analysis. There cannot be a generalized approach that applies everywhere!

kgube reviewed (approved):
LGTM
